dev/NavigationView/NavigationView_ApiTests/NavigationViewTests.cs

```diff
@@ -55,6 +57,95 @@ private NavigationView SetupNavigationView(NavigationViewPaneDisplayMode paneDis
+                navView.MenuItems.Add(new NavigationViewItem() { Content = "Item #9", Icon = new SymbolIcon(Symbol.Cut) });
+                navView.MenuItems.Add(new NavigationViewItem() { Content = "Item #10", Icon = new SymbolIcon(Symbol.Cut) });
+                navView.MenuItems.Add(new NavigationViewItemHeader() { Content = "Item #11" });
+                navView.MenuItems.Add(new NavigationViewItem() { Content = "Item #12", Icon = new SymbolIcon(Symbol.Cut) });
+                navView.MenuItems.Add(new NavigationViewItem() { Content = "Item #13", Icon = new SymbolIcon(Symbol.Cut) });
+                navView.MenuItems.Add(new NavigationViewItemSeparator());
+                navView.MenuItems.Add(new NavigationViewItem() { Content = "Item #15", Icon = new SymbolIcon(Symbol.Cut) });
+
+                navView.IsBackButtonVisible = NavigationViewBackButtonVisible.Visible;
+                navView.IsSettingsVisible = true;
+                navView.PaneDisplayMode = paneDisplayMode;
+                navView.OpenPaneLength = 120.0;
+                navView.ExpandedModeThresholdWidth = 600.0;
+                navView.CompactModeThresholdWidth = 400.0;
+                navView.Width = 800.0;
+                navView.Height = 600.0;
+                navView.Content = "This test should have enough NavigationViewItems to scroll.";
+                MUXControlsTestApp.App.TestContentRoot = navView;
+            });
+
+            IdleSynchronizer.Wait();
+            return navView;
+        }
+
+        [TestMethod]
+        public void VerifyVisualTree()
+        {
+            bool failed = false;
+
+            // Generate a basic NavigationView master file for all the pane display modes
+            foreach (var paneDisplayMode in Enum.GetValues(typeof(NavigationViewPaneDisplayMode)))
+            {
+                var filePrefix = "NavigationView" + paneDisplayMode;
+                NavigationViewPaneDisplayMode displayMode = (NavigationViewPaneDisplayMode)paneDisplayMode;
+
+                // We can skip generating a master file for Left mode since Auto is achieving the same result.
+                if(displayMode == NavigationViewPaneDisplayMode.Left)
+                {
+                    continue;
+                }
+
+                try
+                {
+                    Log.Comment($"Verify visual tree for NavigationViewPaneDisplayMode: {paneDisplayMode}");
+                    var navigationView = SetupNavigationView(displayMode);
+                    VisualTreeTestHelper.VerifyVisualTree(root: navigationView, masterFilePrefix: filePrefix);
+                }
+                catch (Exception e)
+                {
+                    failed = true;
+                    Log.Error(e.Message);
+                }
+
+            }
+
+            // Generate a master file for a special case NavigationView with more items and item variety
+            try
```

Comment:
I don't think you need try catch here?

Reply:
I think grabbing a bunch of master files in one test case is not a unique scenario. @ojhad can you think of ways to reformulate the API to not fail the whole test immediately but instead fail at the end?

Maybe `using(visualverifier) { take a bunch of tree snapshots }` kind of a thing and then the visualverifier class in its Dispose() can throw an exception with a summary of all the comparisons that failed?

Reply:
I think TAEF supports a verification mode where failing 'Verify' call do not throw and instead just log the failure and continue. This might be the most appropriate approach here. This test could enable the non-throwing Verification mode, and so it would churn through all of the verifications while still failing the test in the end.

---
In reply to: [338189647](https://github.com/microsoft/microsoft-ui-xaml/pull/1479#discussion_r338189647) [](ancestors = 338189647)

Reply:
Updated the MSTestInterop implementation to have a DisableVerifyFailureExceptions implementation analogous to the TAEF mode. Also implemented `VisualTreeVerifier` so that the actual visual tree test implementations don't have to worry about checking for each individual failure.